docs/src/userguide/extension_types.rst

```diff
@@ -155,64 +283,120 @@ For known builtin or extension types, Cython translates these into a
-extension type and its :keyword:`cdef` attributes and methods.
+extension type and its C-level attributes (stored in object’s C struct.) and
+:keyword:`cdef`/``@cfunc`` methods.
 
 .. _extension_types_and_none:
 
 Extension types and None
 =========================
 
-When you declare a parameter or C variable as being of an extension type,
-Cython will allow it to take on the value ``None`` as well as values of its
+Cython handles :keyword:`None` values differently in Cython language and when annotations are used.
+
+In the Cython language, when you declare a parameter or C variable as being of an extension type,
+Cython language will allow it to take on the value :keyword:`None` as well as values of its
 declared type. This is analogous to the way a C pointer can take on the value
 ``NULL``, and you need to exercise the same caution because of it. There is no
 problem as long as you are performing Python operations on it, because full
 dynamic type checking will be applied. However, when you access C attributes
 of an extension type (as in the widen_shrubbery function above), it's up to
-you to make sure the reference you're using is not ``None`` -- in the
+you to make sure the reference you're using is not :keyword:`None` -- in the
 interests of efficiency, Cython does not check this.
 
-You need to be particularly careful when exposing Python functions which take
-extension types as arguments. If we wanted to make :func:`widen_shrubbery` a
-Python function, for example, if we simply wrote::
+When annotations are used, the behaviour is close to `PEP-484 <https://www.python.org/dev/peps/pep-0484/>`_.
+The value :keyword:`None` is not allowed when variable is annotated only by type. To allow also :keyword:`None` value,
+``typing.Optional`` must be used.
+
+In the Cython language, you need to be particularly careful when exposing Python functions which take
+on the other hand, annotations are safe since you need to explicitly enable :keyword:`None` value.
+extension types as arguments:
+
+.. tabs::
+
+    .. group-tab:: Pure Python
+
+        .. code-block:: python
 
-    def widen_shrubbery(Shrubbery sh, extra_width): # This is
-        sh.width = sh.width + extra_width           # dangerous!
+            from typing import Optional
 
-then users of our module could crash it by passing ``None`` for the ``sh``
+            def widen_shrubbery(sh: Optional[Shrubbery], extra_width): # This is
+                sh.width = sh.width + extra_width                      # dangerous!
+
+    .. group-tab:: Cython
+
+        .. code-block:: cython
+
+            def widen_shrubbery(Shrubbery sh, extra_width): # This is
+                sh.width = sh.width + extra_width           # dangerous!
+
+The users of our module could crash it by passing :keyword:`None` for the ``sh``
 parameter.
 
-One way to fix this would be::
+In the Cython language, one way to fix this would be to check :keyword:`None` value:
```

Comment:
I partially agree. What I consider challenging when writing documentation is lack of stable defined terms. Cython is complex language with a lot of constructs with different semantics - `cdef`/`@cfunc` functions, `cpdef`/`@ccall` functions, .... My personal opinion is that each term should be documented in glossary and documentation should have links to glossary when the term is used (at least for the first time).

When we will have these terms it will be much easier to write documentation (and even write more clear code in cython itself since we can use this terms in classes/functions etc...). E.g. in documentation we have this horrific sentence:

> C Functions declared using cdef or the `@cfunc` decorator with a Python object return type, like Python functions, will return a None value when execution leaves the function body without an explicit return value [1]

The cdef keyword was already defined above but since we did not name the special functions declared by `cdef` we need to describe (mention that the function is declared by cdef or `@cfunc` decorator) it every time. This is cumbersome and difficult to read. Instead, if during definition of cdef keyword name the construct name is defined (e.g. as *C-function* as an example) explicitly, we can reference it only by the name without mentioning the same all the time. Hence the sentence above will look as follows:

> *C-Functions*  with a Python object return type, like Python functions, will return a None value when execution leaves the function body without an explicit return value

See that I have removed the detail about `cdef`/`@cfunc` since they are not needed. They were introduced during definition of term `C-function` and user can easily go back to term definition (or to glossary) and look up how c-function is declared in Cython.

[1] https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#python-functions-vs-c-functions

Reply:
We used to call them "cdef functions" because that's how they were defined and it was thus obvious to users what was meant. That was before we had "@cfunc". But it was also clearer because "C function" is ambiguous. Is it a function from an external library implemented in C? That is somewhat different from a "cdef function" implemented in Cython, even though in the end both are C functions. Depending on the context, this can be obvious to users or confusing. "cdef function" was a very concrete and clear term. Talking about "Cython implemented C functions" seems fairly verbose in contrast.

I think, nowadays, "C function" is the term that we should use. Maybe not in all cases, but at least when it does not matter whether the function is implemented in Cython on in C. And it should be in the glossary, also mentioning the different names that people would find and the different ways to write it. Here's a proposal, up for debate:

```
cdef function – A Cython implemented -> C function

C function – A function with C interface and calling semantics, either implemented directly in Cython (sometimes referred to as "cdef function"), or implemented in plain C and declared as "extern" for its usage in Cython code.
Cython code uses either the "cdef"  keyword or the "@cfunc" decorator when implementing such a function.
```

Reply:
My point was mainly not about a name itself (name can be C function, cdef function, whatever we decide). My mainly point was that we should define the names and put it in glossary with links to it. It will make documentation much easier to work with.